### PR TITLE
Behavioral Cloning Pytorch

### PR DESCRIPTION
**Pull request by andrewcoh**
### Types of change(s)

- [ ] Bug fix
- [ ] New feature
- [ ] Code refactor
- [ ] Breaking change
- [ ] Documentation update
- [ ] Other (please describe)

### Checklist
- [ ] Added tests that prove my fix is effective or that my feature works
- [ ] Updated the [changelog](https://github.com/Unity-Technologies/ml-agents/blob/master/com.unity.ml-agents/CHANGELOG.md) (if applicable)
- [ ] Updated the [documentation](https://github.com/Unity-Technologies/ml-agents/tree/master/docs) (if applicable)
- [ ] Updated the [migration guide](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Migrating.md) (if applicable)

### CODE REVIEW

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+import numpy as np
+import torch
+
+from mlagents.trainers.policy.torch_policy import TorchPolicy
+from mlagents.trainers.demo_loader import demo_to_buffer
+from mlagents.trainers.settings import BehavioralCloningSettings
+from mlagents.trainers.torch.utils import ModelUtils
+
+
+class TorchBCModule:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:11]: This will probably be renamed after we deprecate TF. 
I would prefer calling this class `BCModule` but put it in a `torch` subfolder (like what @ervteng  is doing with the `network.py` classes)

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        :param strength: The proportion of learning rate used to update through BC.
+        :param steps: The number of steps to anneal BC training over. 0 for continuous training.
+        :param demo_path: The path to the demonstration file.
+        :param batch_size: The batch size to use during BC training.
+        :param num_epoch: Number of epochs to train for during each update.
+        :param samples_per_update: Maximum number of samples to train on during each BC update.
+        """
+        self.policy = policy
+        self.current_lr = policy_learning_rate * settings.strength
+        params = list(self.policy.actor_critic.parameters())
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:36]: Will policy always have a public actor_critic? I wonder if we should have a `parameters` method on `TorchPolicy` to make this more easy to understand.

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+        self.optimizer = torch.optim.Adam(params, lr=self.current_lr)
+
+        _, self.demonstration_buffer = demo_to_buffer(
+            settings.demo_path, policy.sequence_length, policy.behavior_spec
+        )
+
+        self.batch_size = (
+            settings.batch_size if settings.batch_size else default_batch_size
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:44]: I think the default batch size be part of the `BehavioralCloningSettings` (to keep all the defaults at the same place)

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 4]
```diff
@@ -0,0 +1,192 @@
...
+        )
+
+        self.has_updated = False
+        self.use_recurrent = self.policy.use_recurrent
+        self.samples_per_update = settings.samples_per_update
+        # self.out_dict = {
+        #    "loss": self.model.loss,
+        #    "update": self.model.update_batch,
+        #    "learning_rate": self.model.annealed_learning_rate,
+        # }
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:60]:
```suggestion
```

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+    def update(self) -> Dict[str, Any]:
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:62]:
```suggestion
    def update(self) -> Dict[str, np.ndarray]:
```

Comment (vincentpierre) [reply ml-agents/mlagents/trainers/components/bc/module_torch.py:62]: Let's avoid `Any` as much as possible

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+                self.policy.actor_critic.network_body.visual_encoders
+            ):
+                vis_ob = ModelUtils.list_to_tensor(
+                    mini_batch_demo["visual_obs%d" % idx]
+                )
+                vis_obs.append(vis_ob)
+        else:
+            vis_obs = []
+
+        selected_actions, all_log_probs, entropies, values, memories = self.policy.sample_actions(
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:176]:
```suggestion
        selected_actions, all_log_probs, _, _, _ = self.policy.sample_actions(
```

File: ml-agents/mlagents/trainers/components/bc/module_torch.py [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+            all_log_probs=True,
+        )
+        bc_loss = self._behavioral_cloning_loss(
+            selected_actions, all_log_probs, expert_actions
+        )
+        self.optimizer.zero_grad()
+        bc_loss.backward()
+
+        self.optimizer.step()
+        run_out = {"loss": bc_loss.detach().cpu().numpy()}
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/components/bc/module_torch.py:191]: I don't think there is a need (yet) to return a dict containing a single value over just returning the loss.

File: ml-agents/mlagents/trainers/components/bc/module_torch.py (hunk 1 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/components/bc/module_torch.py:11]: I agree. What of your reward providers?

File: ml-agents/mlagents/trainers/components/bc/module_torch.py (hunk 7 above)
Comment (andrewcoh) [reply ml-agents/mlagents/trainers/components/bc/module_torch.py]: When the decaying learning rate is merged that will also be passed here

File: ml-agents/mlagents/trainers/torch/components/bc/module.py [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+        :param strength: The proportion of learning rate used to update through BC.
+        :param steps: The number of steps to anneal BC training over. 0 for continuous training.
+        :param demo_path: The path to the demonstration file.
+        :param batch_size: The batch size to use during BC training.
+        :param num_epoch: Number of epochs to train for during each update.
+        :param samples_per_update: Maximum number of samples to train on during each BC update.
+        """
+        self.policy = policy
+        self.current_lr = policy_learning_rate * settings.strength
+        params = list(self.policy.actor_critic.parameters())
```
Comment (vincentpierre) [inline ml-agents/mlagents/trainers/torch/components/bc/module.py:34]: No need to cast to list, the Adam optimizer can take the dict directly

File: ml-agents/mlagents/trainers/torch/components/bc/module.py [hunk 9]
```diff
@@ -147,14 +147,11 @@ def _update_batch(
                 )
             )
 
-        memories = [
-            ModelUtils.list_to_tensor(mini_batch_demo["memory"][i])
-            for i in range(
-                0, len(mini_batch_demo["memory"]), self.policy.sequence_length
+        memories = []
+        if self.policy.actor_critic.use_lstm:
```
Comment (ervteng) [inline ml-agents/mlagents/trainers/torch/components/bc/module.py:151]: You can also use `policy.use_recurrent`, probably more generalizable than calling the `actor_critic`
